homeassistant/components/google_translate/tts.py

```diff
@@ -96,23 +96,245 @@
+    "co.in",
+    "iq",
+    "is",
+    "it",
+    "je",
+    "com.jm",
+    "jo",
+    "co.jp",
+    "co.ke",
+    "com.kh",
+    "ki",
+    "kg",
+    "co.kr",
+    "com.kw",
+    "kz",
+    "la",
+    "com.lb",
+    "li",
+    "lk",
+    "co.ls",
+    "lt",
+    "lu",
+    "lv",
+    "com.ly",
+    "co.ma",
+    "md",
+    "me",
+    "mg",
+    "mk",
+    "ml",
+    "com.mm",
+    "mn",
+    "ms",
+    "com.mt",
+    "mu",
+    "mv",
+    "mw",
+    "com.mx",
+    "com.my",
+    "co.mz",
+    "com.na",
+    "com.ng",
+    "com.ni",
+    "ne",
+    "nl",
+    "no",
+    "com.np",
+    "nr",
+    "nu",
+    "co.nz",
+    "com.om",
+    "com.pa",
+    "com.pe",
+    "com.pg",
+    "com.ph",
+    "com.pk",
+    "pl",
+    "pn",
+    "com.pr",
+    "ps",
+    "pt",
+    "com.py",
+    "com.qa",
+    "ro",
+    "ru",
+    "rw",
+    "com.sa",
+    "com.sb",
+    "sc",
+    "se",
+    "com.sg",
+    "sh",
+    "si",
+    "sk",
+    "com.sl",
+    "sn",
+    "so",
+    "sm",
+    "sr",
+    "st",
+    "com.sv",
+    "td",
+    "tg",
+    "co.th",
+    "com.tj",
+    "tl",
+    "tm",
+    "tn",
+    "to",
+    "com.tr",
+    "tt",
+    "com.tw",
+    "co.tz",
+    "com.ua",
+    "co.ug",
+    "co.uk",
+    "com.uy",
+    "co.uz",
+    "com.vc",
+    "co.ve",
+    "vg",
+    "co.vi",
+    "com.vn",
+    "vu",
+    "ws",
+    "rs",
+    "co.za",
+    "co.zm",
+    "co.zw",
+    "cat",
+]
+
+DEFAULT_TLD = "com"
+
+MAP_LANG_TLD = {
+    "en-us": {"lang": "en", "tld": "com"},
+    "en-gb": {"lang": "en", "tld": "co.uk"},
+    "en-uk": {"lang": "en", "tld": "co.uk"},
+    "en-au": {"lang": "en", "tld": "com.au"},
+    "en-ca": {"lang": "en", "tld": "ca"},
+    "en-in": {"lang": "en", "tld": "co.in"},
+    "en-ie": {"lang": "en", "tld": "ie"},
+    "en-za": {"lang": "en", "tld": "co.za"},
+    "fr-ca": {"lang": "fr", "tld": "ca"},
+    "fr-fr": {"lang": "fr", "tld": "fr"},
+    "pt-br": {"lang": "pt", "tld": "com.br"},
+    "pt-pt": {"lang": "pt", "tld": "pt"},
+    "es-es": {"lang": "es", "tld": "es"},
+    "es-us": {"lang": "es", "tld": "com"},
+}
+
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {vol.Optional(CONF_LANG, default=DEFAULT_LANG): vol.In(SUPPORT_LANGUAGES)}
+    {
+        vol.Optional(CONF_LANG, default=DEFAULT_LANG): vol.In(SUPPORT_LANGUAGES),
+        vol.Optional("tld", default=DEFAULT_TLD): vol.In(SUPPORT_TLD),
+    }
 )
 
 
 async def async_get_engine(hass, config, discovery_info=None):
     """Set up Google speech component."""
-    return GoogleProvider(hass, config[CONF_LANG])
+    return GoogleProvider(hass, config[CONF_LANG], config["tld"])
 
 
 class GoogleProvider(Provider):
     """The Google speech API provider."""
 
-    def __init__(self, hass, lang):
+    def __init__(self, hass, lang, tld):
         """Init Google TTS service."""
         self.hass = hass
-        self._lang = lang
+        if lang in MAP_LANG_TLD:
+            self._lang = MAP_LANG_TLD[lang]["lang"]
+            self._tld = MAP_LANG_TLD[lang]["tld"]
```

Comment:
This is fine.
If you want to, you can help the type checker by specifying the type of `MAP_LANG_TLD`:

```py
MAP_LANG_TLD: dict[str, dict[str, str]] = {
```

Or stricter:
```py
@dataclasses.dataclass
class Dialect:
    """Language and TLD for a dialect supported by Google Translate."""

    lang: str
    tld: str


MAP_LANG_TLD: dict[str, Dialect] = {
...
}

...

self._lang = MAP_LANG_TLD[lang].lang
````

Reply:
The dataclass was a neat way to make this work. Took me a bit to understand how to get it converted from the dict. Thank you @emontnemery!